### ISSUE TITLE
dead_code false positive for const used in definition of tuple struct

### ISSUE THREAD

**Issue by crumblingstatue** [rust-lang/rust#56281]
```Rust
const LEN: usize = 4;

#[derive(Debug)]
struct Wrapper([u8; LEN]);

fn main() {
    println!("{:?}", Wrapper([0, 1, 2, 3]));
}

```

```
warning: constant item is never used: `LEN`
 --> src/main.rs:1:1
  |
1 | const LEN: usize = 4;
  | ^^^^^^^^^^^^^^^^^^^^^
```

If the struct is a record struct with a named field, the false positive doesn't trigger.

**Comment by oli-obk** [comment]
I have not been able to get anywhere with this issue.

I did some experiments in https://github.com/rust-lang/rust/blob/master/src/librustc/middle/dead.rs and I believe the issue is that using a tuple struct's constructor will only mark the constructor as used (which has a different DefId and NodeId than the tuple struct item). This guess has a further datapoint than my reading: Adding `let x: Wrapper;` will silence the dead code warning about `LEN`.

The reason that we are not getting a dead code warning about `Wrapper` is that checking the item for "having been used" will also accept just the constructor having been used: https://github.com/rust-lang/rust/blob/041254b81495a5aa67af839e00b890e78ed0cbeb/src/librustc/middle/dead.rs#L467-L468 

I have not been able to figure out a way to go from a constructor back to its item, there only seems to be code for the other way around.

We might be able to hack it by getting the function signature of the item and checking the return type, but in order to do that, we'd need to know for sure that it's a tuple struct constructor.

Alternatively we can just build up a map from constructors to items by filling it in whenever we encounter a tuple struct item.

ok enough rubber-ducking. I'm gonna try the sentence above now.
